pyo3-macros-backend/src/pyclass.rs

```diff
@@ -1185,34 +1186,30 @@ fn impl_complex_enum_variant_cls(
 }
 
 fn impl_complex_enum_variant_match_args(
-    ctx: &Ctx,
+    ctx @ Ctx { pyo3_path, .. }: &Ctx,
     variant_cls_type: &syn::Type,
     field_names: &mut Vec<Ident>,
-) -> (MethodAndMethodDef, syn::ImplItemConst) {
+) -> syn::Result<(MethodAndMethodDef, syn::ImplItemFn)> {
     let ident = format_ident!("__match_args__");
-    let match_args_const_impl: syn::ImplItemConst = {
-        let args_tp = field_names.iter().map(|_| {
-            quote! { &'static str }
-        });
+    let mut match_args_impl: syn::ImplItemFn = {
         parse_quote! {
-            #[allow(non_upper_case_globals)]
-            const #ident: ( #(#args_tp,)* ) = (
-                #(stringify!(#field_names),)*
-            );
+            #[classattr]
+            fn #ident(py: #pyo3_path::Python<'_>) -> #pyo3_path::PyResult<#pyo3_path::Bound<'_, #pyo3_path::types::PyTuple>> {
+                #pyo3_path::types::PyTuple::new::<&str, _>(py, [
+                    #(stringify!(#field_names),)*
+                ])
+            }
         }
     };
 
-    let spec = ConstSpec {
-        rust_ident: ident,
-        attributes: ConstAttributes {
-            is_class_attr: true,
-            name: None,
-        },
-    };
-
-    let variant_match_args = gen_py_const(variant_cls_type, &spec, ctx);
+    let spec = FnSpec::parse(
+        &mut match_args_impl.sig,
+        &mut match_args_impl.attrs,
+        Default::default(),
+    )?;
+    let variant_match_args = impl_py_class_attribute(variant_cls_type, &spec, ctx)?;
```

Comment:
I had an initial feeling that building the syntax tree and then parsing it might be a bit convoluted (we could just construct a `FnSpec` directly) but maybe this is actually quite elegant because it means the `#[classattr]` defined above fully defines the final result.

It would have been nice to avoid the fallibility, but I think it's fine on balance.

Reply:
> it means the `#[classattr]` defined above fully defines the final result.

This was my motivation for this. I think it's more robust against accidental inconsistencies (now and in the future)

> It would have been nice to avoid the fallibility, but I think it's fine on balance.

Yeah, I was a bit sad as well . It _should_ never fail, so we could also `expect`, but that did not feel much better to me.

Happy New Year to you as well 🎆